azure-iot-device/azure/iot/device/common/auth/sastoken.py

```diff
@@ -94,3 +94,49 @@ def _build_token(self):
     def expiry_time(self):
         """Expiry Time is READ ONLY"""
         return self._expiry_time
+
+
+class NonRenewableSasToken(object):
+    def __init__(self, sastoken_string):
+        self._token = sastoken_string
+        self._token_info = get_sastoken_info_from_string(self._token)
+
+    def __str__(self):
+        return self._token
+
+    @property
+    def expiry_time(self):
+        """Expiry Time is READ ONLY"""
+        return int(self._token_info["se"])
+
+    @property
+    def resource_uri(self):
+        "Resource URI is READ ONLY"
+        uri = self._token_info["sr"]
+        return urllib.parse.unquote(uri)
+
+
+REQUIRED_SASTOKEN_FIELDS = ["sr", "sig", "se"]
+VALID_SASTOKEN_FIELDS = REQUIRED_SASTOKEN_FIELDS + ["skn"]
+
+
+def get_sastoken_info_from_string(sastoken_string):
+    pieces = sastoken_string.split("SharedAccessSignature ")
+    if len(pieces) != 2:
+        raise SasTokenError("Invalid SasToken string: Not a SasToken ")
+
+    # Get sastoken info as dictionary
+    try:
+        sastoken_info = dict(map(str.strip, sub.split("=", 1)) for sub in pieces[1].split("&"))
```

Comment:
>dict(map(str.strip, sub.split("=", 1)) [](start = 24, length = 38)

Wow!  Impressive.  I tried to write something similar yesterday, but you definitely win.

Reply:
Had to do it one line

---
In reply to: [506681080](https://github.com/Azure/azure-iot-sdk-python/pull/692#discussion_r506681080) [](ancestors = 506681080)